content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+```
+$ wasm2wat ./target/wasm32-unknown-unknown/release/my_project.wasm
+```
+
+This command will convert a WebAssembly binary to WAT and output it to stdout.
+
+```wasm
+(module
+  (table (;0;) 1 1 funcref)
+  (memory (;0;) 16)
+  (global $__stack_pointer (mut i32) (i32.const 1048576))
+  (global (;1;) i32 (i32.const 1048576))
+  (global (;2;) i32 (i32.const 1048576))
+  (export "memory" (memory 0))
+  (export "__data_end" (global 1))
+  (export "__heap_base" (global 2)))
+```
+
+It is with outrage that we discover that our `add` function has been completely removed from the binary. All we are left with is a stack pointer, and two globals designating where the data section ends and the heap starts. Turns out declaring a function as `pub` is _not_ enough to get it to show up in our final WebAssembly module. I kinda wish it was enough, but I suspect `pub` is exclusive about Rust module visibility, not about linker-level visibility.
+
+The quickest way to change the compiler's behavior is to explicitly give the function an export name:
+
+|||codediff|rust
++ #[export_name = "add"]
+  pub fn add(left: usize, right: usize) -> usize {
+      left + right
+  }
+|||
+
+If you don’t intend to change the functions external name, you can also use `#[no_mangle]` instead, which will have the same effect. I feel like `no_mangle` hides the true intention of the developer here, so I tend to prefer the `export_name`. 
+
+Having given our `add` function an export name, we can compile the project again and inspect the resulting WebAssembly file:
+
+|||codediff|wasm
+  (module
++   (type (;0;) (func (param i32 i32) (result i32)))
++   (func $add (type 0) (param i32 i32) (result i32)
++     local.get 1
++     local.get 0
++     i32.add)
+    (table (;0;) 1 1 funcref)
+    (memory (;0;) 16)
+    (global $__stack_pointer (mut i32) (i32.const 1048576))
+    (global (;1;) i32 (i32.const 1048576))
+    (global (;2;) i32 (i32.const 1048576))
+    (export "memory" (memory 0))
++   (export "add" (func $add))
+    (export "__data_end" (global 1))
+    (export "__heap_base" (global 2)))
+|||
+
+This module is easy to run in Node, or Deno or even the browser:
+
+```js
+// Node
+const data = require("fs").readFileSync("./my_project.wasm");
+// Deno
+const data = await Deno.readFile("./my_project.wasm");
+// Web
+const data = await fetch("./my_project.wasm").then(r => r.arrayBuffer());
+
+const importObj = {
+};
+const {instance} = await WebAssembly.instantiate(data, importObj);
+instance.exports.add(40, 2) // returns 42
+```
+
+And suddenly, we have pretty much all the power of Rust at our fingertips to write WebAssembly. 
+
+> **Streaming:** On the web it is advisable to _not_ call `r.arrayBuffer()` and instead pass the `Response` straight to `WebAssembly.instantiateStreaming`. This allows the WebAssembly compiler to start compiling while the file is being downloaded.
```

Comment:
I'd rather just inline this into the `// Web` example above, rather than showing the discouraged version first and the recommended one in a side-note.